src/lib/storage/dictionary_segment/dictionary_encoder.hpp

```diff
@@ -28,112 +29,92 @@ class DictionaryEncoder : public SegmentEncoder<DictionaryEncoder<Encoding>> {
-      return _encode_dictionary_segment(pmr_vector<T>{values.cbegin(), values.cend(), values.get_allocator()},
-                                        value_segment);
+      return _encode_dictionary_segment(dictionary, values, null_values, allocator);
     }
   }
 
  private:
   template <typename U, typename T>
   static ValueID _get_value_id(const U& dictionary, const T& value) {
     return ValueID{static_cast<ValueID::base_type>(
-        std::distance(dictionary.cbegin(), std::lower_bound(dictionary.cbegin(), dictionary.cend(), value)))};
+        std::distance(dictionary->cbegin(), std::lower_bound(dictionary->cbegin(), dictionary->cend(), value)))};
   }
 
   template <typename U, typename T>
-  std::shared_ptr<BaseEncodedSegment> _encode_dictionary_segment(
-      U dictionary, const std::shared_ptr<const ValueSegment<T>>& value_segment) {
-    const auto& values = value_segment->values();
-    const auto alloc = values.get_allocator();
-
-    // Remove null values from value vector
-    if (value_segment->is_nullable()) {
-      const auto& null_values = value_segment->null_values();
-
-      // Swap values to back if value is null
-      auto erase_from_here_it = dictionary.end();
-      auto null_it = null_values.crbegin();
-      for (auto dict_it = dictionary.rbegin(); dict_it != dictionary.rend(); ++dict_it, ++null_it) {
-        if (*null_it) {
-          std::iter_swap(dict_it, --erase_from_here_it);
-        }
-      }
-
-      // Erase null values
-      dictionary.erase(erase_from_here_it, dictionary.end());
-    }
-
-    std::sort(dictionary.begin(), dictionary.end());
-    dictionary.erase(std::unique(dictionary.begin(), dictionary.end()), dictionary.end());
-    dictionary.shrink_to_fit();
-
-    auto attribute_vector = pmr_vector<uint32_t>{values.get_allocator()};
-    attribute_vector.reserve(values.size());
-
-    const auto null_value_id = static_cast<uint32_t>(dictionary.size());
-
-    if (value_segment->is_nullable()) {
-      const auto& null_values = value_segment->null_values();
-
-      /**
-       * Iterators are used because values and null_values are of
-       * type tbb::concurrent_vector and thus index-based access isn’t in O(1)
-       */
-      auto value_it = values.cbegin();
-      auto null_value_it = null_values.cbegin();
-      for (; value_it != values.cend(); ++value_it, ++null_value_it) {
-        if (*null_value_it) {
-          attribute_vector.push_back(null_value_id);
-          continue;
-        }
-
-        const auto value_id = _get_value_id(dictionary, *value_it);
-        attribute_vector.push_back(value_id);
-      }
-    } else {
-      auto value_it = values.cbegin();
-      for (; value_it != values.cend(); ++value_it) {
-        const auto value_id = _get_value_id(dictionary, *value_it);
-        attribute_vector.push_back(value_id);
+  std::shared_ptr<BaseEncodedSegment> _encode_dictionary_segment(const U& dictionary, const std::vector<T>& values,
```

Comment:
This only creates the attribute vector, but it does not create a dictionary segment from scratch. In itself, it seems useless. Move it into a lambda `create_attribute_vector` that is part of `_on_encode`?

Reply:
I adapted the function in a way that is is only responsible for the creation of compressed attribute vectors.
However, I am too stupid to move it into a simple lambda since I fail to get around the error, that dictionary template type cannot be deduced (_on_encode had two template parameters, T and S, and S was the parameter of the lambda function, S could not be deduced).